Closes: #18535 - Skip incompatible plugins during startup

bctiemann opened the pull request:
### Closes: #18535

Changes the behavior of the `PluginConfig.validate` method so that if a plugin with incompatible min or max version specified, it raises an `IncompatiblePluginError` which is caught in `settings.py` and registering the plugin is skipped, allowing NetBox to launch cleanly with the plugin disabled.

bctiemann committed:
Skip incompatible plugins during startup and remove from PLUGINS

bctiemann committed:
Skip incompatible plugins during startup and remove from PLUGINS

bctiemann commented on netbox/netbox/settings.py:
```diff
@@ -813,6 +815,16 @@ def _setting(name, default=None):
+    try:
+        plugin_config.validate(PLUGINS_CONFIG[plugin_name], RELEASE.version)
+    except IncompatiblePluginError as e:
+        print(f'Unable to load plugin {plugin_name}: {e}')
```
I'm not sure about adding a `print` here, as it would be the first of its kind in this file. Is there a better way to emit this warning?

a084ed22 commented:
Just to understand, this would print a warning at netbox upgrade time? The experience of figuring out which plugin has a strict max_version specified and breaks the upgrade process even on patch upgrades during the database migration step could really use to be more explanatory.

bctiemann commented:
Yes, it would look like this:

```
Unable to load plugin netbox_branching: Plugin netbox_branching requires NetBox maximum version 4.2.1 (current: 4.2.2).
Watching for file changes with StatReloader
Performing system checks...
```

bctiemann committed:
Handle exceptions on request processors in incompatible plugins, and display status in Plugins page

bctiemann committed:
Revert "Handle exceptions on request processors in incompatible plugins, and display status in Plugins page"

This reverts commit d97bf2ab146114cc13d751878a17a383de0fd5f8.

bctiemann committed:
Resolve merge conflicts

jeremystretch commented on netbox/core/plugins.py:
```diff
@@ -91,15 +91,15 @@ def get_local_plugins(plugins=None):
         if k in plugins:
             plugins[k].is_local = True
-            plugins[k].is_installed = True
+            plugins[k].is_installed = k in registry['plugins']['installed']
```
I think we can copy this over directly?
```suggestion
            plugins[k].is_installed = v.is_installed
```

jeremystretch reviewed (changes requested):
We might want to rename the table columns. Currently, a plugin which fails to load shows false under "Installed" but confusingly still reports an installed version:

![screenshot](https://github.com/user-attachments/assets/d474be1b-635a-4449-8f87-5720d29ab1ab)

We should also highlight any plugins which failed to load to draw attention to them.

bctiemann committed:
Rename Installed column to Active, and add custom PluginActiveColumn with tooltip

bctiemann committed:
Fix is_installed

bctiemann commented:
I renamed "Installed" to "Active", and I added a custom table column that allows an alert icon with a tooltip that explains the incompatibility issue:

<img width="1091" alt="Screenshot 2025-02-23 at 6 51 43 PM" src="https://github.com/user-attachments/assets/1c904510-1135-4635-bfa0-c9d3049b3903" />
<img width="1085" alt="Screenshot 2025-02-23 at 6 50 32 PM" src="https://github.com/user-attachments/assets/b1467ad7-ea83-489f-b492-3ca2010334e5" />

jeremystretch commented on netbox/netbox/settings.py:
```diff
@@ -811,6 +812,15 @@ def _setting(name, default=None):
         )
 
+    # Validate version compatibility and user-provided configuration settings and assign defaults
+    if plugin_name not in PLUGINS_CONFIG:
+        PLUGINS_CONFIG[plugin_name] = {}
+    try:
+        plugin_config.validate(PLUGINS_CONFIG[plugin_name], RELEASE.version)
```
We can simplify this a bit:
```suggestion
    try:
        plugin_config.validate(PLUGINS_CONFIG.get(plugin_name, {}), RELEASE.version)
```

bctiemann replied:
Actually I don't think this can be done this way. Trying to `validate` in one fell swoop like this calls `get_plugin_config` which raises an `ImproperlyConfigured` error which I'm not trapping in this try-except (intentionally), and that fires before it checks whether the key is in the dict. It's weird and ugly but the way I had it is the only way to do it without changing `get_plugin_config` too with even more abstractions.

jeremystretch reviewed (changes requested):
This renders an alert for every non-installed plugin. The value for `is_installed` should probably be None if the plugin has not been added to `PLUGINS`, and False if it has been added but failed to load.

![screenshot](https://github.com/user-attachments/assets/4bf906dd-9293-4fcd-8be6-dd49e5a02c13)

bctiemann committed:
Simplify plugin_config.validate syntax

Co-authored-by: Jeremy Stretch <jstretch@netboxlabs.com>

bctiemann committed:
Merge feature

bctiemann committed:
Revert "Merge feature"

This reverts commit d1ea60f08270b9e79d30b9fa9859049aa371f4c6.

bctiemann committed:
Undo simplification

bctiemann committed:
Add failed_to_load logic

jeremystretch commented on netbox/netbox/tables/columns.py:
```diff
@@ -707,3 +707,22 @@ class DistanceColumn(TemplateColumn):
         super().__init__(template_code=template_code, order_by=order_by, **kwargs)
+
+
+class PluginActiveColumn(BooleanColumn):
```
Could you just use a TemplateColumn then? I'd like to avoid introducing a new column class that only has one use case.

bctiemann committed:
Use a TemplateColumn for is_installed

bctiemann committed:
Remove custom column class

bctiemann committed:
Remove merge vestige

jeremystretch commented on netbox/core/plugins.py:
```diff
@@ -66,7 +66,10 @@ class Plugin:
     release_recent_history: list[PluginVersion] = field(default_factory=list)
     is_local: bool = False  # extra field for locally installed plugins
     is_installed: bool = False
+    failed_to_load: bool = False
```
I'm having a hard time discerning the difference between `is_local` and `is_installed`. It seems that any installed plugin has `is_local` set to True. Additionally, `failed_to_load` feels backward to me.

Could we simplify these to two attributes: `is_enabled` and `is_installed` (or `is_loaded`)? These would indicate whether the plugin is listed in `settings.PLUGINS` and whether is has been successfully loaded into the registry, respectively.

bctiemann replied:
How about this -- `is_local` (which mirrors the `local_plugins` etc verbage, as opposed to `catalog_plugins`); and `is_loaded`, which indicates whether it's loaded into the registry. This removes `is_installed` as redundant, and the `is_installed` column remains on the table class but is purely rendered via the template logic.

jeremystretch commented on netbox/core/tables/template_code.py:
```diff
@@ -14,3 +14,13 @@
 """
+
+PLUGIN_IS_INSTALLED = """
+{% if record.failed_to_load %}
```
A plugin could fail to load for other reasons though, presumably. I'd avoid assuming it was due to a version incompatibility.

bctiemann replied:
Rephrased the tooltip accordingly. It's possible to do an on-the-fly validation of the version range to see whether it should be suggested at all, but that got a bit too hairy.

bctiemann committed:
Simplify plugin attributes for is_installed column

jeremystretch committed:
Use placeholders for false values to increase legibility of the plugins table
